runtime/compiler/env/DependencyTable.cpp

```diff
@@ -0,0 +1,232 @@
+/*******************************************************************************
+ * Copyright IBM Corp. and others 2024
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] https://openjdk.org/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+ *******************************************************************************/
+
+#if !defined(PERSISTENT_COLLECTIONS_UNSUPPORTED)
+
+#include "control/CompilationThread.hpp"
+#include "env/DependencyTable.hpp"
+#include "env/J9SharedCache.hpp"
+#include "env/PersistentCHTable.hpp"
+
+TR_AOTDependencyTable::TR_AOTDependencyTable(TR_J9SharedCache *sharedCache) :
+   _isActive(true),
+   _sharedCache(sharedCache),
+   _tableMonitor(TR::Monitor::create("JIT-AOTDependencyTableMonitor")),
+   _offsetMap(decltype(_offsetMap)::allocator_type(TR::Compiler->persistentAllocator()))
+   { }
+
+void
+TR_AOTDependencyTable::classLoadEvent(TR_OpaqueClassBlock *clazz, bool isClassLoad, bool isClassInitialization)
+   {
+   auto ramClass = (J9Class *)clazz;
+
+   uintptr_t classOffset = TR_SharedCache::INVALID_ROM_CLASS_OFFSET;
+   if (!_sharedCache->isClassInSharedCache(clazz, &classOffset))
+      return;
+
+   // We only need to check if clazz matches its cached version on load; on
+   // initialization, it will be in the _offsetMap if it did match.
+   if (isClassLoad && !_sharedCache->classMatchesCachedVersion(ramClass, NULL))
+      return;
+
+   OMR::CriticalSection cs(_tableMonitor);
+   if (!isActive())
+      return;
+
+   try
+      {
+      classLoadEventAtOffset(ramClass, classOffset, isClassLoad, isClassInitialization);
+      }
+   catch (std::exception&)
+      {
+      deactivateTable();
+      }
+   }
+
+void
+TR_AOTDependencyTable::classLoadEventAtOffset(J9Class *ramClass, uintptr_t offset, bool isClassLoad, bool isClassInitialization)
+   {
+   auto entry = getOffsetEntry(offset, isClassLoad);
+   TR_ASSERT(entry || !isClassLoad, "Class %p offset %lu initialized without loading");
+
+   if (isClassLoad)
```

Comment:
I guess code will evolve, because right now, if it's not a class load there is no point in calling getOffsetEntry(offset, isClassLoad);